File: SimSim/Menus.m
//
//  Menus.m
//  SimSim
//
//  Created by Daniil Smelov on 05.04.18.
//  Copyright © 2018 Daniil Smelov. All rights reserved.
//

#import "Menus.h"
#import "Actions.h"
#import "Realm.h"
#import "Settings.h"
#import "Tools.h"
#import "CommanderOne.h"

@implementation Menus

#define ACTION_ICON_SIZE 16

// TODO: make it less hardcoded :)

#define FINDER_ICON_PATH @"/System/Library/CoreServices/Finder.app"
#define TERMINAL_ICON_PATH @"/Applications/Utilities/Terminal.app"
#define ITERM_ICON_PATH @"/Applications/iTerm.app"
#define CMDONE_ICON_PATH @"/Applications/Commander One.app"

//----------------------------------------------------------------------------
+ (NSNumber*) addAction:(NSString*)title
              toSubmenu:(NSMenu*)submenu
                forPath:(NSString*)path
               withIcon:(NSString*)iconPath
              andHotkey:(NSNumber*)hotkey
                   does:(nullable SEL)selector
{
    NSMenuItem* item = [[NSMenuItem alloc] initWithTitle:title
                                                  action:selector
                                           keyEquivalent:[hotkey stringValue]];
    [item setTarget:[Actions class]];
    [item setRepresentedObject:path];
    
    NSImage* icon = [[NSWorkspace sharedWorkspace] iconForFile:iconPath];
    [icon setSize: NSMakeSize(ACTION_ICON_SIZE, ACTION_ICON_SIZE)];
    [item setImage:icon];
    
    [submenu addItem:item];
    
    return @([hotkey intValue] + 1);
}

//----------------------------------------------------------------------------
+ (NSNumber*) addAction:(NSString*)title
              toSubmenu:(NSMenu*)submenu
                forPath:(NSString*)path
             withHotkey:(NSNumber*)hotkey
                   does:(nullable SEL)selector
{
    NSMenuItem* item = [[NSMenuItem alloc] initWithTitle:title
                                                  action:selector
                                           keyEquivalent:[hotkey stringValue]];
    
    [item setTarget:[Actions class]];
    
    [item setRepresentedObject:path];
    [submenu addItem:item];
    
    return @([hotkey intValue] + 1);
}

//----------------------------------------------------------------------------
+ (Realm*) realmModule
{
    static Realm * realm = nil;
    
    if (realm == nil)
        realm = [Realm new];
    
    return realm;
}

//----------------------------------------------------------------------------
+ (NSNumber*) addActionForRealmTo:(NSMenu*)menu
                          forPath:(NSString*)path
                       withHotkey:(NSNumber*)hotkey
{
    if ([Realm isRealmAvailableForPath:path])
    {
        NSImage* icon = [[NSWorkspace sharedWorkspace] iconForFile:[Realm applicationPath]];
        [icon setSize: NSMakeSize(ACTION_ICON_SIZE, ACTION_ICON_SIZE)];
        
        [[self realmModule] generateRealmMenuForPath:path forMenu:menu withHotKey:hotkey icon:icon];
        
        return @([hotkey intValue] + 1);
    }
    
    return hotkey;
}

//----------------------------------------------------------------------------
+ (NSNumber*) addActionForiTermTo:(NSMenu*)menu
                          forPath:(NSString*)path
                       withHotkey:(NSNumber*)hotkey
{
    CFStringRef iTermBundleID = CFStringCreateWithCString(CFAllocatorGetDefault(), "com.googlecode.iterm2", kCFStringEncodingUTF8);
    CFArrayRef iTermAppURLs = LSCopyApplicationURLsForBundleIdentifier(iTermBundleID, NULL);
    
    if (iTermAppURLs)
    {
        hotkey = [self addAction:@"iTerm" toSubmenu:menu forPath:path
                        withIcon:ITERM_ICON_PATH andHotkey:hotkey
                            does:@selector(openIniTerm:)];
        
        CFRelease(iTermAppURLs);
        return @([hotkey intValue] + 1);
    }
    
    CFRelease(iTermBundleID);

    return hotkey;
}

//----------------------------------------------------------------------------
+ (void) addSubMenusToItem:(NSMenuItem*)item usingPath:(NSString*)path
{
    NSMenu* subMenu = [NSMenu new];
    
    NSNumber* hotkey = @1;
    
    hotkey = [self addAction:@"Finder" toSubmenu:subMenu forPath:path
                    withIcon:FINDER_ICON_PATH andHotkey:hotkey
                        does:@selector(openInFinder:)];
    
    hotkey = [self addAction:@"Terminal" toSubmenu:subMenu forPath:path
                    withIcon:TERMINAL_ICON_PATH andHotkey:hotkey
                        does:@selector(openInTerminal:)];

    hotkey = [self addActionForRealmTo:subMenu forPath:path withHotkey:hotkey];
    hotkey = [self addActionForiTermTo:subMenu forPath:path withHotkey:hotkey];
    
    if ([CommanderOne isCommanderOneAvailable])
    {
        hotkey = [self addAction:@"Commander One" toSubmenu:subMenu forPath:path
                        withIcon:CMDONE_ICON_PATH andHotkey:hotkey
                            does:@selector(openInCommanderOne:)];
    }
    
    [subMenu addItem:[NSMenuItem separatorItem]];
    
    hotkey = [self addAction:@"Copy path to Clipboard" toSubmenu:subMenu forPath:path
                  withHotkey:hotkey
                        does:@selector(copyToPasteboard:)];
    
    if ([Tools simulatorRunning])
    {
        hotkey = [self addAction:@"Take Screenshot" toSubmenu:subMenu forPath:path
                      withHotkey:hotkey
                            does:@selector(takeScreenshot:)];
    }
    
    hotkey = [self addAction:@"Reset application data" toSubmenu:subMenu forPath:path
                  withHotkey:hotkey
                        does:@selector(resetApplication:)];
    
    [item setSubmenu:subMenu];
}

//----------------------------------------------------------------------------
+ (void) addApplication:(Application*)application toMenu:(NSMenu*)menu
{
    NSString* title =
    [NSString stringWithFormat:@"%@ (%@)", application.bundleName, application.version];
    
    // This path will be opened on click
    NSString* applicationContentPath = application.contentPath;
    
    NSMenuItem* item = [[NSMenuItem alloc] initWithTitle:title action:@selector(openInWithModifier:) keyEquivalent:@"\0"];
    [item setTarget:[Actions class]];

    [item setRepresentedObject:applicationContentPath];
    [item setImage:application.icon];
    
    [self addSubMenusToItem:item usingPath:applicationContentPath];
    
    [menu addItem:item];
}

//----------------------------------------------------------------------------
+ (void) addApplications:(NSArray<Application*>*)installedApplicationsData toMenu:(NSMenu*)menu
{
    for (NSUInteger i = 0; i < [installedApplicationsData count]; i++)
    {
        Application* application = installedApplicationsData[i];
        
        [self addApplication:application toMenu:menu];
    }
}

//----------------------------------------------------------------------------
+ (void) addServiceItemsToMenu:(NSMenu*)menu
{
    NSMenuItem* startAtLogin =
    [[NSMenuItem alloc] initWithTitle:@"Start at Login" action:@selector(handleStartAtLogin:) keyEquivalent:@""];
    
    BOOL isStartAtLoginEnabled = [Settings isStartAtLoginEnabled];
    
    [startAtLogin setState: isStartAtLoginEnabled ? NSOnState : NSOffState];
    
    [startAtLogin setRepresentedObject:@(isStartAtLoginEnabled)];
    [menu addItem:startAtLogin];
    
    NSString* appVersion = [NSString stringWithFormat:@"About %@ %@", [[NSRunningApplication currentApplication] localizedName], [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleShortVersionString"]];
    NSMenuItem* about = [[NSMenuItem alloc] initWithTitle:appVersion action:@selector(aboutApp:) keyEquivalent:@"I"];
    [menu addItem:about];
    
    NSMenuItem* quit = [[NSMenuItem alloc] initWithTitle:@"Quit" action:@selector(exitApp:) keyEquivalent:@"Q"];
    [menu addItem:quit];
}

#define MAX_RECENT_SIMULATORS 5

//----------------------------------------------------------------------------
+ (NSMenu*) createApplicationMenuAt:(NSStatusItem*)statusItem
{
    NSMenu* menu = [NSMenu new];
    
    NSMutableArray* simulators = [Tools activeSimulators];
    
    NSArray* recentSimulators = [simulators sortedArrayUsingComparator:^NSComparisonResult(id a, id b)
    {
        NSDate* l = [(Simulator*)a date];
        NSDate* r = [(Simulator*)b date];
        return [r compare:l];
    }];
    
    int simulatorsCount = 0;
    for (Simulator* simulator in recentSimulators)
    {
        NSArray<Application*>* installedApplications = [Tools installedAppsOnSimulator:simulator];
        
        if ([installedApplications count])
        {
            NSString* simulator_title = [NSString stringWithFormat:@"%@ (%@)",
                                         simulator.name,
                                         simulator.os];
            
            NSMenuItem* simulatorMenuItem = [[NSMenuItem alloc] initWithTitle:simulator_title action:nil keyEquivalent:@""];
            [simulatorMenuItem setEnabled:NO];
            [menu addItem:simulatorMenuItem];
            [self addApplications:installedApplications toMenu:menu];
            
            simulatorsCount++;
            if (simulatorsCount >= MAX_RECENT_SIMULATORS)
                break;
        }
    }
    
    [menu addItem:[NSMenuItem separatorItem]];
    
    [self addServiceItemsToMenu:menu];
    
    return menu;
}


@end

File: SimSim/Menus.swift
//
//  Menus.swift
//  SimSim
//
//  Created by Daniil Smelov on 13/04/2018.
//  Copyright © 2018 Daniil Smelov. All rights reserved.
//

import Foundation
import Cocoa

//============================================================================
class Menus: NSObject
{
    private static var realm = Realm()
    
    //----------------------------------------------------------------------------
    class func addAction(_ title: String, toSubmenu submenu: NSMenu,
                         forPath path: String, withIcon iconPath: String,
                         andHotkey hotkey: NSNumber,
                         does selector: Selector) -> NSNumber
    {
        let item = NSMenuItem(title: title, action: selector, keyEquivalent: hotkey.stringValue)
        item.target = Actions.self
        item.representedObject = path
        
        item.image = NSWorkspace.shared.icon(forFile: iconPath)
        item.image?.size = NSMakeSize(CGFloat(Constants.iconSize), CGFloat(Constants.iconSize))
        
        submenu.addItem(item)
        
        return NSNumber(value: hotkey.intValue + 1)
    }
    
    //----------------------------------------------------------------------------
    class func addAction(_ title: String, toSubmenu submenu: NSMenu,
                               forPath path: String, withHotkey hotkey: NSNumber,
                               does selector: Selector) -> NSNumber
    {
        let item = NSMenuItem(title: title, action: selector, keyEquivalent: hotkey.stringValue)

        item.target = Actions.self
        item.representedObject = path
        
        submenu.addItem(item)
        return NSNumber(value: hotkey.intValue + 1)
    }

    //----------------------------------------------------------------------------
    class func realmModule() -> Realm
    {
        return realm
    }

    //----------------------------------------------------------------------------
    class func addActionForRealm(to menu: NSMenu, forPath path: String,
                                 withHotkey hotkey: NSNumber) -> NSNumber
    {
        guard Realm.isRealmAvailable(forPath: path) else
        {
            return hotkey
        }
        
        let icon = NSWorkspace.shared.icon(forFile: Realm.applicationPath())
        icon.size = NSMakeSize(CGFloat(Constants.iconSize), CGFloat(Constants.iconSize))
        Realm.generateRealmMenu(forPath: path, for: menu, withHotKey: hotkey, icon: icon)
        return NSNumber(value: hotkey.intValue + 1)
    }

    //----------------------------------------------------------------------------
    class func addActionForiTerm(to menu: NSMenu, forPath path: String, withHotkey hotkey: NSNumber) -> NSNumber
    {
        let iTermAppURLs = LSCopyApplicationURLsForBundleIdentifier(Constants.Other.iTermBundle as CFString, nil)
        
        guard iTermAppURLs != nil else {
            return hotkey
        }
        
        let newkey = addAction(Constants.Actions.iTerm, toSubmenu: menu, forPath: path, withIcon: Constants.Paths.iTermApp,
                               andHotkey: hotkey, does: #selector(Actions.openIniTerm(_:)))
        
        return NSNumber(value: newkey.intValue + 1)
    }
    
    //----------------------------------------------------------------------------
    class func addSubMenus(to item: NSMenuItem, usingPath path: String)
    {
        let subMenu = NSMenu()
        var hotkey = NSNumber(value: 1)
        
        
        hotkey = addAction(Constants.Actions.finder, toSubmenu: subMenu, forPath: path, withIcon: Constants.Paths.finderApp, andHotkey: hotkey, does: #selector(Actions.open(inFinder:)))
        
        hotkey = addAction(Constants.Actions.terminal, toSubmenu: subMenu, forPath: path, withIcon: Constants.Paths.terminalApp, andHotkey: hotkey, does: #selector(Actions.open(inTerminal:)))

        hotkey = addActionForRealm(to: subMenu, forPath: path, withHotkey: hotkey)
        hotkey = addActionForiTerm(to: subMenu, forPath: path, withHotkey: hotkey)
        
        if Tools.commanderOneAvailable()
        {
            hotkey = addAction(Constants.Actions.commanderOne, toSubmenu: subMenu, forPath: path, withIcon: Constants.Paths.commanderOneApp, andHotkey: hotkey, does: #selector(Actions.open(inCommanderOne:)))
        }

        subMenu.addItem(NSMenuItem.separator())
        hotkey = addAction(Constants.Actions.clipboard, toSubmenu: subMenu, forPath: path, withHotkey: hotkey, does: #selector(Actions.copy(toPasteboard:)))
        
        hotkey = addAction(Constants.Actions.reset, toSubmenu: subMenu, forPath: path, withHotkey: hotkey, does: #selector(Actions.resetApplication(_:)))
        item.submenu = subMenu
    }
    
    //----------------------------------------------------------------------------
    class func add(_ application: Application, to menu: NSMenu)
    {
        let title = "\(application.bundleName ?? "nil") \(application.version ?? "nil")"
        // This path will be opened on click
        let applicationContentPath = application.contentPath
        let item = NSMenuItem(title: title, action: #selector(Actions.openIn(withModifier:)), keyEquivalent: "\0")
        item.target = Actions.self
        item.representedObject = applicationContentPath
        item.image = application.icon
        self.addSubMenus(to: item, usingPath: applicationContentPath)
        menu.addItem(item)
    }

    //----------------------------------------------------------------------------
    class func add(_ applications: [Application], to menu: NSMenu)
    {
        for application in applications
        {
            add(application, to: menu)
        }
    }

    //----------------------------------------------------------------------------
    class func add(appGroup: AppGroup, to menu: NSMenu)
    {
        let item = NSMenuItem(title: appGroup.identifier, action: #selector(Actions.openIn(withModifier:)), keyEquivalent: "\0")
        item.target = Actions.self
        item.representedObject = appGroup.path

        self.addSubMenus(to: item, usingPath: appGroup.path)

        menu.addItem(item)
    }
    
    //----------------------------------------------------------------------------
    class func add(appExtension: AppExtension, to menu: NSMenu)
    {
        let item = NSMenuItem(title: appExtension.identifier, action: #selector(Actions.openIn(withModifier:)), keyEquivalent: "\0")
        item.target = Actions.self
        item.representedObject = appExtension.path
        
        self.addSubMenus(to: item, usingPath: appExtension.path)
        
        menu.addItem(item)
    }

    //----------------------------------------------------------------------------
    class func addServiceItems(to menu: NSMenu)
    {
        /*
         Start at login API was changed, so this feature needs to be updated:
         https://jogendra.dev/implementing-launch-at-login-feature-in-macos-apps

         let startAtLogin = NSMenuItem(title: Constants.Actions.login, action: #selector(Actions.handleStart(atLogin:)), keyEquivalent: "")
         startAtLogin.target = Actions.self
         let isStartAtLoginEnabled = Settings.isStartAtLoginEnabled

         startAtLogin.state = isStartAtLoginEnabled ? .on : .off
         startAtLogin.representedObject = isStartAtLoginEnabled
         menu.addItem(startAtLogin)
         */

        let appVersion = "About \(NSRunningApplication.current.localizedName ?? "") \(Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "")"
        
        let about = NSMenuItem(title: appVersion, action: #selector(Actions.aboutApp(_:)), keyEquivalent: "I")
        about.target = Actions.self
        menu.addItem(about)
        
        let quit = NSMenuItem(title: Constants.Actions.quit, action: #selector(Actions.exitApp(_:)), keyEquivalent: "Q")
        quit.target = Actions.self
        menu.addItem(quit)
    }
    
    //----------------------------------------------------------------------------
    class func updateApplicationMenu(_ menu: NSMenu, at statusItem: NSStatusItem) -> Void
    {
        menu.removeAllItems()

        let simulators = Tools.activeSimulators()
        
        if simulators.isEmpty
        {
            let noSimulatorsItem = NSMenuItem(title: "No Simulators", action: nil, keyEquivalent: "")
            noSimulatorsItem.isEnabled = false
            menu.addItem(noSimulatorsItem)
        }

        let recentSimulators = simulators
            .sorted { $0.date > $1.date }
            .prefix(Constants.maxRecentSimulators)

        for simulator in recentSimulators
        {
            let installedApplications = Tools.installedApps(on: simulator)
            let sharedAppGroups = Tools.sharedAppGroups(on: simulator)
            let appExtensions = Tools.appExtensions(on: simulator)
            
            guard installedApplications.count != 0 else
            {
                continue
            }
            
            let simulatorTitle = simulator.name + " " + simulator.os
            let simulatorMenuItem = NSMenuItem(title: simulatorTitle, action: nil, keyEquivalent: "")
            simulatorMenuItem.isEnabled = false
            menu.addItem(simulatorMenuItem)
            add(installedApplications, to: menu)
            sharedAppGroups.forEach { add(appGroup: $0, to: menu) }
            appExtensions.forEach { add(appExtension: $0, to: menu) }
        }
        menu.addItem(NSMenuItem.separator())
        addServiceItems(to: menu)
    }
}


























